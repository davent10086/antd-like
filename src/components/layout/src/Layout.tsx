import React, { createContext} from 'react';
import type { LayoutProps } from '../interface';
import classNames from 'classnames';
import '../style/index.scss';

/**
 * LayoutContext用于在Layout组件和其子组件之间传递侧边栏相关的操作方法
 */
export const LayoutContext = createContext({
  siderHook: {
    addSider: () => {},
    removeSider: () => {}
  }
});

/**
 * Layout组件是页面布局的基础容器组件
 * @param props - Layout组件的属性
 * @param props.direction - 布局方向，可选值为 'horizontal' | 'vertical'，默认为 'horizontal'
 * @param props.children - 布局内的子元素
 * @param props.className - 自定义CSS类名
 * @param props.hasSider - 是否包含侧边栏
 * @returns 返回一个布局容器DOM元素
 */
const Layout: React.FC<LayoutProps> = (props) => {
  const { 
    direction = 'horizontal', 
    children, 
    className, 
    hasSider,
    ...restProps 
  } = props;

  const prefixCls = 'ant-layout';

  // 构建组件的CSS类名
  const classes = classNames(
    prefixCls,
    {
      [`${prefixCls}-has-sider`]: typeof hasSider === 'boolean' ? hasSider : false,
      [`${prefixCls}-vertical`]: direction === 'vertical',
    },
    className,
  );

  return (
    <LayoutContext.Provider
      value={{
        siderHook: {
          addSider: () => {},
          removeSider: () => {}
        }
      }}
    >
      <div className={classes} {...restProps}>
        {children}
      </div>
    </LayoutContext.Provider>
  );
};

export default Layout;