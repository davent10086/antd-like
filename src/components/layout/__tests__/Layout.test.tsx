import { render } from '@testing-library/react';
import Layout from '../index';

const { Header, Content, Footer, Sider } = Layout;

describe('Layout', () => {
  it('should render correctly', () => {
    const { container } = render(
      <Layout>
        <Header>Header</Header>
        <Content>Content</Content>
        <Footer>Footer</Footer>
      </Layout>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render correctly with sider', () => {
    const { container } = render(
      <Layout hasSider>
        <Sider>Sider</Sider>
        <Layout>
          <Header>Header</Header>
          <Content>Content</Content>
          <Footer>Footer</Footer>
        </Layout>
      </Layout>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should apply vertical class when direction is vertical', () => {
    const { container } = render(
      <Layout direction="vertical">
        <Header>Header</Header>
        <Content>Content</Content>
      </Layout>
    );
    expect(container.querySelector('.ant-layout-vertical')).toBeInTheDocument();
  });

  it('should apply has-sider class when hasSider is true', () => {
    const { container } = render(
      <Layout hasSider>
        <Sider>Sider</Sider>
      </Layout>
    );
    expect(container.querySelector('.ant-layout-has-sider')).toBeInTheDocument();
  });

  it('should support custom className', () => {
    const { container } = render(
      <Layout className="custom-class">
        <Content>Content</Content>
      </Layout>
    );
    expect(container.querySelector('.custom-class')).toBeInTheDocument();
  });

  it('should support custom style', () => {
    const { container } = render(
      <Layout style={{ backgroundColor: 'red' }}>
        <Content>Content</Content>
      </Layout>
    );
    expect(container.firstChild).toHaveStyle('background-color: red');
  });
});