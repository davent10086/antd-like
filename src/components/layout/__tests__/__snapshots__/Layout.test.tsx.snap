// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout should render correctly 1`] = `
<div
  class="ant-layout"
>
  <header
    class="ant-layout-header"
    style="height: 64px; line-height: 64px;"
  >
    Header
  </header>
  <main
    class="ant-layout-content"
  >
    Content
  </main>
  <footer
    class="ant-layout-footer"
    style="height: 64px; line-height: 64px;"
  >
    Footer
  </footer>
</div>
`;

exports[`Layout should render correctly with sider 1`] = `
<div
  class="ant-layout ant-layout-has-sider"
>
  <aside
    class="ant-layout-sider"
    style="flex: 0 0 200px; max-width: 200px; min-width: 200px; width: 200px;"
  >
    <div
      class="ant-layout-sider-children"
    >
      Sider
    </div>
  </aside>
  <div
    class="ant-layout"
  >
    <header
      class="ant-layout-header"
      style="height: 64px; line-height: 64px;"
    >
      Header
    </header>
    <main
      class="ant-layout-content"
    >
      Content
    </main>
    <footer
      class="ant-layout-footer"
      style="height: 64px; line-height: 64px;"
    >
      Footer
    </footer>
  </div>
</div>
`;
